File: day-03-arrays-vs-linked-lists/python/linked-list.py
def create_node(value): 
    return { "value": value, "next": None }

def prepend(head, value):
    new_node = create_node(value)
    new_node["next"] = head
    return new_node

def traverse(head):
    current = head
    while current is not None:
        print(current["value"])
        current = current["next"]

def find(head, value):
    current = head
    while current is not None:
        if current["value"] == value:
            return current
        current = current["next"]
    return None

def insert_after(node, value):
    new_node = create_node(value)
    new_node["next"] = node["next"]
    node["next"] = new_node

# Test
head = create_node(10)
print(head)
print("\n===================\n")

head = prepend(head, 5)
print(head)
print("\n===================\n")

head = prepend(head, 3)
print(head)
print("\n===================\n")

traverse(head)
print("\n===================\n")

found = find(head, 5)
if found:
    insert_after(found, 7)
print(head)
print("\n===================\n")

traverse(head)

File: day-03-arrays-vs-linked-lists/js/linked-list.js
// Create node
function createNode(value) {
  return { value: value, next: null };
}

// Add a value to the front, return the new head
function prepend(head, value) {
  let newNode = createNode(value);
  newNode.next = head;
  return newNode;
}

// Prints all values from head to end
function traverse(head) {
  let current = head;
  while (current !== null) {
    console.log(current.value);
    current = current.next;
  }
}

// Find node with given value, return node or null
function find(head, value) {
  let current = head;
  while (current !== null) {
    if (current.value === value) {
      return current;
    }
    current = current.next;
  }
  return null;
}

// Insert a new value after the given node
function insertAfter(node, value) {
  let newNode = createNode(value);
  newNode.next = node.next;
  node.next = newNode;
}

// Test
let head = createNode(10);
console.log(head);
console.log("\n===================\n");

head = prepend(head, 5);
console.log(head);
console.log("\n===================\n");

head = prepend(head, 3);
console.log(head);
console.log("\n===================\n");

traverse(head);
console.log("\n===================\n");

let found = find(head, 5);

if (found) {
  insertAfter(found, 7);
}
traverse(head);
